solutions/usaco-574.mdx

```diff
@@ -52,60 +52,95 @@ int main() {
 Instead of a knapsack solution, a BFS/DFS can be done to iterate though **all** attainable states and simply choose the one with maximum fullness. The scene can be thought of as a graph by defining each edge to be the transition from fullness $i$ to $j$. Each state is defined by fullness and whether Bessie has drank water. Since there are $2T$ states and each state is visited at most once, the time and space complexity of this solution is $\mathcal{O}(T)$. Below is a BFS implementation.
 
 <LanguageSection>
-
 <CPPSection>
 
 ```cpp
-#include <bits/stdc++.h>
-using namespace std;
-
-const int maxn = 5e6 + 5;
-bool pass[maxn][2];
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <set>
+#include <queue>
 
-struct node {
-	int val;
-	bool drunk;
-};
+using std::cout;
+using std::endl;
+using std::vector;
+using std::pair;
 
-int bfs(int t, int a, int b) {
-	int res = INT_MIN;
-	queue<node> q;
-	q.push({0, 0});
-
-	while (!q.empty()) {
-		int value = q.front().val;
-		bool drunk = q.front().drunk;
-		q.pop();
-		res = max(res, value);
-
-		if(value + a <= t && !pass[value + a][drunk]) {
-			q.push({value + a, drunk});
-			pass[value + a][drunk] = true;
-		}
-
-		if(value + b <= t && !pass[value + b][drunk]) {
-			q.push({value + b, drunk});
-			pass[value + b][drunk] = true;
+int main() {
+	std::ifstream read("feast.in");
+	int max_fullness;
+	int orange;
+	int lemon;
+	read >> max_fullness >> orange >> lemon;
+
+	std::set<pair<int, bool>> poss_fullness{{0, false}};
+	std::queue<pair<int, bool>> queue;
+	queue.push({0, false});
+	while (!queue.empty()) {
+		auto [curr, drank_water] = queue.front();
+		queue.pop();
+
+		// Drink water (if we haven't drank it already)
+		if (!drank_water) {
+			pair<int, bool> water{curr / 2, true};
+			if (!poss_fullness.count(water)) {
+				queue.push(water);
+				poss_fullness.insert(water);
+			}
 		}
 
-		if(!drunk && !pass[value / 2][1]) {
-			q.push({value / 2, 1});
-			pass[value / 2][1] = true;
+		// Eat a fruit
+		for (int fruit : vector<int>{orange, lemon}) {
+			int after = curr + fruit;
+			pair<int, bool> to_add{after, drank_water};
+			if (after <= max_fullness && !poss_fullness.count(to_add)) {
+				queue.push(to_add);
+				poss_fullness.insert(to_add);
+			}
 		}
 	}
-	return res;
-}
 
-int main() {
-	freopen("feast.in", "r", stdin);
-	freopen("feast.out", "w", stdout);
+	int best = 0;
+	for (const auto& [fullness, _] : poss_fullness) {
+		best = std::max(best, fullness);
+	}
 
-	int t, a, b;
-	cin >> t >> a >> b;
-	cout << bfs(t, a, b);
+	std::ofstream("feast.out") << best << endl;
 }
 ```
 
 </CPPSection>
+<PySection>
+
+```py
+from collections import deque
+	
+with open('feast.in') as read:
+	max_fullness, orange, lemon = map(int, read.readline().split())
+
+fullness = {(0, False)}
```

Comment:
this is just a visited array. "fullness" as its name is misleading

Reply:
yeah, it represents all the possible fullness levels

Reply:
just a suggestion: make the variable of the set called "visited" and add a comment that says "all possible fullnesses with or without water." It wasn't clear to me

Reply:
renamed it to `poss_fullness`